packages/create-amplify/src/create_amplify.ts

```diff
@@ -7,23 +7,59 @@
   If customers have a cached version of the create-amplify package, they might execute that cached version even after we publish features and fixes to the package on npm.
  */
 
-import { NpmPackageManagerController } from './npm_package_manager_controller.js';
 import { ProjectRootValidator } from './project_root_validator.js';
 import { AmplifyProjectCreator } from './amplify_project_creator.js';
+import {
+  PackageManagerBase,
+  type PackageManagerName,
+  type PackageManagerProps,
+  packageManagerControllerFactory,
+} from './package-manager-controller/index.js';
+import { projectInitializerFactory } from './project-initializer/index.js';
 import { InitialProjectFileGenerator } from './initial_project_file_generator.js';
-import { NpmProjectInitializer } from './npm_project_initializer.js';
 import { getProjectRoot } from './get_project_root.js';
 import { GitIgnoreInitializer } from './gitignore_initializer.js';
+import { logger } from './logger.js';
 
 const projectRoot = await getProjectRoot();
 
+const getPackageManager: () => PackageManagerProps = () => {
+  const packageManager = new PackageManagerBase();
+  if (!process.env.npm_config_user_agent) {
+    logger.warn('Could not determine package manager, defaulting to npm');
+    return packageManager.getPackageManager('npm');
+  }
+
+  const userAgent = process.env.npm_config_user_agent;
+  const packageManagerAndVersion = userAgent.split(' ')[0];
+  const packageManagerName = packageManagerAndVersion.split('/')[0];
+
+  if (packageManagerName === 'yarn') {
+    const yarnMajorVersion = packageManagerAndVersion
+      .split('/')[1]
+      .split('.')[0];
+    const yarnName: PackageManagerName = `${packageManagerName}-${
+      yarnMajorVersion === '1' ? 'classic' : 'modern'
+    }`;
+    return packageManager.getPackageManager(yarnName);
+  }
+  return packageManager.getPackageManager(
```

Comment:
shouldn't this be `packageManagerFactory.getPackageManager(packageManagerName)` ?

Or even better `packageManagerFactory.getPackageManager(userAgent)` - so that all logic that determines which package manager controller/initializer we want to create.

Also we should probably have ``packageManagerFactory.getPackageManagerController(...)` and ``packageManagerFactory.getProjectInitializer(...)` (unless we merge them as suggested in other comment).

Reply:
Done 11796e3a5